docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,214 @@
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+      - [Examples](#examples)
+        - [Alpha -> Beta Graduation](#alpha---beta-graduation)
+        - [Beta -> GA Graduation](#beta---ga-graduation)
+        - [Removing a deprecated flag](#removing-a-deprecated-flag)
+    - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Glossary
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+Enable opt in automated health checking and remediation of unhealthy nodes backed by machines.
+
+## Motivation
+- Reduce administrative overhead to run a cluster.
+
+- Ability to respond to failures of machines and keep the cluster nodes healthy.
+
+### Goals
+- Enable automated remediation for pools of machines/nodes (e.g a machineSet)
+
+- Allow to define different unhealthy criterias based on node conditions for different pools.
+
+- Providing administrator-defined short-circuiting of automated remediation when multiple nodes are unhealthy at the same time.
+
+### Non-Goals/Future Work
+- Recording long-term stable history of all health-check failures or remediations.
+
+- Providing a mechanisim to guarantee that application quorum for N members is maintained at any time.
+
+## Proposal
+The machine health checker (MHC) does a best effort to keep nodes healthy in the cluster, by removing machines backing unhealthy Nodes.
+
+It provides a short-circuit mechanism and limits remediation when the `maxUnhealthy` threshold is reached for a targeted group of machines. This is similar to what the node life cycle controller does for reducing the eviction rate as nodes become unhealthy in a given zone. E.g a large number of nodes in a single zone are down due to a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+### Unhealthy criteria:
+A machine is unhealthy when:
+
+- The referenced node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has a nodeRef but the referenced node is not found.
+- The Machine is in phase "Failed", It has `ErrorMessage`/`ErrorReason` set.
+
+If any of those criteria are met for longer than the given timeouts and the `maxUnhealthy` threshold has not been reached yet, remediation is triggered.
```

Comment:
It is not clear whether having more than `maxUnhealthy` nodes results in either limiting remediation to `maxUnhealthy` nodes or stopping remediation entirely.

If I define "node is older than 30 days" as an unhealthy node condition criterion, it is likely for the threshold to be exceeded. In that case, the desired behavior would be to rate-limit remediation.

Reply:
`maxUnhealthy` will completely short-circuit remediation. I'm totally on-board for rate-limiting but unless there're strong objections I'd prefer to keep this proposal intentionally simple and follow up with a RFE.

Reply:
```suggestion
If any of those criteria are met for longer than the given timeouts and the `maxUnhealthy` threshold has not been reached yet, remediation is triggered. If more than `maxUnhealthy` Machines are unhealthy then none will be remediated. 
```